src/engine.ml

```diff
@@ -448,8 +447,9 @@ let incoming_control_client config rng session channel now op data =
               Some (highest, highest)
             else
               Some (v, highest)
+        and peer_name = Config.find Verify_x509_name config
         in
-        Tls.(Engine.client (Config.client ?ciphers ?version ~certificates ~authenticator ()))
+        Tls.(Engine.client (Config.client ?ciphers ?version ?peer_name ~certificates ~authenticator ()))
```

Comment:
The default type seems to be to check `CN`. I think this will check against the subjectAltName? See `--x509-username-field` in `man openvpn`. I think we would require `x509-username-field ext:subjectAltName` in order to do this.

I am also thinking if this is something we should rather pass to the authenticator. I don't remember what the ` ~peer_name` argument does.

Reply:
This is true, there's a discrepancy from OpenVPN and our implementation. But I think ours is more intuitive.

OpenVPN seems to only use the name of the subject of the certificate (esp. the Common Name fragment). We use the set of common name and SubjectAlternativeName.

I'm happy to document this behaviour somewhere. I'm also fine to have this discrepancy.

Reply:
The `~peer_name` specifies which hostname is expected on the server side. I.e. the `validate_chain` uses the peer_name for verifying the name in the X.509 certificate. It is also used in the ClientHello as SNI (Server Name Identification).

The "authenticator" is a function that gets passed the `?ip` and `?hostname` -- by the TLS stack.

Reply:
I documented that in [6bbb95a](https://github.com/robur-coop/miragevpn/pull/89/commits/6bbb95ae225afdef84b5cb43c21613cc88fe600f)

Reply:
Hmm, ok. Yes, I think we should document it if we diverge.

I wrote this monstrosity if we want to support the openvpn behavior
```ocaml
        let authenticator =
          Option.fold (Config.find Verify_x509_name config)
            ~none:authenticator
            ~some:(fun peer_name ?ip ~host certs ->
                Result.bind
                  (authenticator ?ip ~host certs)
                  (fun r ->
                     match certs with
                     | [] -> Error `EmptyCertificateChain
                     | peer :: _ ->
                       let dn = X509.Certificate.subject peer in
                       if Option.equal String.equal (Some peer_name)
                           (X509.Distinguished_name.common_name dn)
                       then
                         Ok r
                       else
                         Error (`Msg "peer name mismatch")))
        in
```

Reply:
We can invent a new `x509-username-field`, say `peer-name`, that does what we do and have that as a default, and then we can implement parsing `x509-username-field` -- and then users can specify `x509-username-field CN` if they really want that.
But yes - for another PR :)